src/server.c

```
@@ -94,22 +137,47 @@ void* handle_client(void* arg) {
     memcpy(key, serverKey.key, AES_KEY_LEN);
     memcpy(iv, serverKey.iv, AES_IV_SIZE);
 
-    // Read the client ID
-    if ((bytes_read = recv(client_socket, client_id, sizeof(client_id), 0)) > 0) {
-        client_id[bytes_read] = '\0';
-        printf("%s connected\n", client_id);
-    }
-
     // Sends the key to the client
     if (send(client_socket, serverKey.key, AES_KEY_LEN, 0) != AES_KEY_LEN) {
         fail_client_with_error("Error sending AES key", client_socket);
     }
-
     // Sends the iv the client
     if (send(client_socket, serverKey.iv, AES_IV_SIZE, 0) != AES_IV_SIZE) {
         fail_client_with_error("Error sending AES IV", client_socket);
     }
+    
+    
+       // Read the client ID(username)
+    if ((bytes_read = recv(client_socket, client_id, sizeof(client_id), 0)) > 0) {
+        client_id[bytes_read] = '\0';    
+    }
+    
+  
+       // Read the client password and authenticate
+    if ((bytes_read = recv(client_socket, password, sizeof(password), 0)) > 0) {
+       password[bytes_read] = '\0';
+        SCPHeader* header = (SCPHeader*)password;
+        uint8_t msg_type = header->msg_type;
 
+        // Decrypt the received payload
+        unsigned char plaintext[BUFFER_SIZE];
+        int ciphertext_len = ntohs(header->payload_length);
+        unsigned char* cipherpassword = (unsigned char*)(password + sizeof(SCPHeader));
+        check = check_credentials(client_id, cipherpassword); 
+
+        
+    }
+    
+     // check if the user is authenticated
+      if(check != 1)
+        {
+          char error[BUFFER_SIZE] = "Fail to authenticate, Check your credentials"; 
+          printf("%s\n", error); 
+          send_message(client_socket,-4, 0, 0, error);
+          close(client_socket);  
```

Should `fail_client_with_error` instead of the printf & close.

---

Th is not cutting connection. Error message when password is wrong. Socket will remain open.

---

`close(client_socket);` closes the socket

---

changed